src/HappyConnOpener.cc

```diff
@@ -435,9 +434,9 @@ HappyConnOpener::status() const
     static SBuf buf;
     buf.clear();
 
-    SBufStream os(buf);
+    PackableStream os(buf);
 
-    os.write(" [", 2);
+    os << " [";
```

Comment:
The PR contains many small changes that go beyond the declared scope. Some of them seem wrong to me. I will go through them one by one eventually, but it will take time.

Reply:
Is this one of them?  If so why the unexpected unique use of write() to send two bytes to a stream?

Reply:
> why the unexpected unique use of write() to send two bytes to a stream

It may eliminate a strlen() call. There are many similar optimizations in Squid. We should be hiding (and securing) them by introducing length-aware constants, and we often do, but this and other status() code needs to be rewritten for other reasons anyway, so it just mimics what several other similar status() methods do.

----

Please note that, ideally, we should not be discussing why we should not be doing out-of-scope changes that require a discussion.

Reply:
Undone. Though your logic can be applied to reverse many optimizations in Squid code on grounds that other code does things inefficiently.  ie. leaving this write() call here is actively preserving the collection of technical debt in Squid.

Reply:
> Amos: your logic can be applied to reverse many optimizations in Squid code on grounds that other code does things inefficiently. ie. leaving this write() call here is actively preserving the collection of technical debt in Squid.

I do not follow. Leaving the official write() call "as is" _preserves_ the existing (alleged) optimization. The PR was _removing_ that optimization. Please clarify how leaving this write() "as is" would make anything _worse_.

Reply:
C++11 assigns fixed-strings like this a special ```&char[N]``` type which can be handled by streambuf::xsputn(char*, int N) when the compiler is building optimized code. Note that the stdlib operator<<(char*) documentation says N is calculated "_as if_ strlen(s) were used" - not that strlen(s) is actually happening at run-time.

Reply:
You are right that an optimizing compiler may use the knowledge of the string literal length to avoid a strlen() call. This fact significantly weakens my simple "eliminates strlen()" example -- strlen() is likely to be avoided in both cases by a good optimizing compiler.

However, there is another (more complex) reason why the official code is an optimization (that cannot be performed by a compiler): The official code uses unformatted output. The PR-attempted operator `<<` replacement uses formatted output. An optimizing compiler cannot (and, in my tests, does not) replace one with the other because that would alter output in cases that a compiler cannot detect. Formatted output code naturally does more before (and after) calling streambuf::sputn() -- the unformatted output function used by formatted and unformatted code paths. Thus, the official code remains an optimization even in compiler-optimized builds.

We can, of course, discuss the merits and drawbacks of this optimization and decide that it should be altered or removed, but all of that is outside this PR scope. This PR is not the right place for such discussions and code changes.

----

Thank you for removing this out of scope change. I am resolving this change request even though I have not checked whether other out-of-scope changes this change request complains about have been removed. If they were not, I will deal with them separately, after we resolve the primary problem with this PR -- SBuf virtualization.